### PR TITLE
Add Merkle tree creation

### PR DESCRIPTION
**Pull request by jm-clius**
This PR is a step towards completing https://github.com/status-im/nim-waku/issues/686

### Background

It adds Merkle tree creation to the existing DNS discovery features.
The `builder`can
- [x] create a Merkle tree from ENR or link entries according to [EIP-1459](https://eips.ethereum.org/EIPS/eip-1459)
- [x] retrieve the subdomain and TXT record for each Merkle tree entry if deployed at a given domain

### Next steps:

- More comprehensive tests, especially for large subtrees
- Add ability to sign tree
- DNS Resolver (https://github.com/status-im/nim-waku/issues/615)
- Create an example tree and deploy to a DNS provider

### CODE REVIEW

File: discovery/dnsdisc/builder.nim [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  ## TXT record representation
+  var txtRecord: string
+
+  let sig = Base64Url.encode(rootEntry.signature)
+  try:
+    txtRecord = fmt"{RootPrefix} e={rootEntry.eroot} l={rootEntry.lroot} seq={rootEntry.seqNo} sig={sig}"
+  except ValueError:
+    return err("Failed to format root entry")
+
+  ok(txtRecord)
```
Comment (staheri14) [inline discovery/dnsdisc/builder.nim]:
```suggestion
  return ok(txtRecord)
```

File: discovery/dnsdisc/builder.nim [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    
+    if subdomainRes.isErr:
+      return err("Failed to build: " & subdomainRes.error)
+    
+    if txtRecordRes.isErr:
+      return err("Failed to build: " & txtRecordRes.error)
+
+    treeRecords[subdomainRes.get() & "." & domain] = txtRecordRes.get()
+  
+  ok(treeRecords)
```
Comment (staheri14) [inline discovery/dnsdisc/builder.nim]:
```suggestion
  return ok(treeRecords)
```

File: discovery/dnsdisc/builder.nim [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  try:
+    txtRecord = subtreeEntry.toTXTRecord().tryGet()
+  except ValueError:
+    return err("Failed to format subtree entry")
+  
+  let
+    keccakHash = keccak256.digest(txtRecord.toBytes()).data[0..15]
+    subdomain = Base32.encode(keccakHash)
+  
+  ok(subdomain)
```
Comment (staheri14) [inline discovery/dnsdisc/builder.nim]:
```suggestion
  return ok(subdomain)
```

File: discovery/dnsdisc/builder.nim [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+  case subtreeEntry.kind:
+    of Enr:
+      txtRecord = subtreeEntry.enrEntry.record.toURI()
+    of Link:
+      txtRecord = LinkPrefix & subtreeEntry.linkEntry.str
+    of Branch:
+      txtRecord = BranchPrefix & subtreeEntry.branchEntry.children.join(",")
+  
+  ok(txtRecord)
```
Comment (staheri14) [inline discovery/dnsdisc/builder.nim]:
```suggestion
  return ok(txtRecord)
```

**Review by staheri14 — APPROVED** [review]
Nice!

File: discovery/dnsdisc/builder.nim [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+  Subtree* = object # A coherent section of a larger tree. Can contain branch and leaf nodes.
+    subtreeRoot*: SubtreeEntry
+    subtreeEntries*: seq[SubtreeEntry]
+
+####################
+# Helper functions #
+####################
+
+proc toTXTRecord*(rootEntry: RootEntry): BuilderResult[string] =
```
Comment (oskarth) [inline discovery/dnsdisc/builder.nim:35]: How come BuilderResult is its own type? Not quite following

**Review by oskarth — APPROVED** [review]

File: discovery/dnsdisc/builder.nim (hunk 5 above)
Comment (jm-clius) [reply discovery/dnsdisc/builder.nim:35]: It is to indicate that `Result`s in the builder all have `string` as an error type. In this case the return type is also a string.
See for example [similar implementation in eth2](https://github.com/status-im/nimbus-eth2/blob/44f652f704f14c67e648f2b14dee652d44b858c5/beacon_chain/spec/keystore.nim#L128) and the [docs for `Result`](https://github.com/status-im/nim-stew/blob/master/stew/results.nim#L25-L29)

File: discovery/dnsdisc/builder.nim (hunk 1 above)
Comment (jm-clius) [reply discovery/dnsdisc/builder.nim]: Thanks! :) Done in 306c2e5

File: discovery/dnsdisc/builder.nim (hunk 2 above)
Comment (jm-clius) [reply discovery/dnsdisc/builder.nim]: Done in 306c2e5

File: discovery/dnsdisc/builder.nim (hunk 4 above)
Comment (jm-clius) [reply discovery/dnsdisc/builder.nim]: Done in 306c2e5

File: discovery/dnsdisc/builder.nim (hunk 3 above)
Comment (jm-clius) [reply discovery/dnsdisc/builder.nim]: Done in 306c2e5

**Pull request merged**